### ISSUE TITLE
Co-locating Errors and Data

### ISSUE THREAD

**Issue by jdlubrano** [rmosolgo/graphql-ruby#3197]
I was reading through the GraphQL Ruby guides regarding errors and error-handling.  I am very interested in whether there are any suggested approaches for encoding errors as data into one's schema.  In theory, I understand it.  For mutations, I understand how it works.  For queries, though, I am wondering whether there is a suggested approach.  I have attempted to provide an example of what I am struggling with below.

So I have a schema with Books and Authors.  As a starting point, this contrived system is meant to oversimplify a Graph API aggregating data from an organization's microservice-driven backend.  It is safe to assume that any Book or Author data would be fetched via a RESTful HTTP to a separate service. I have set up Books and Authors to be resolved as objects that I have named (probably terribly) `DistributedModel`.  These `DistributedModel`s provide an interface (i.e. an `id` method) so that the objects can be used to construct RESTful API calls.

```ruby
module Types
  class QueryType < Types::BaseObject
    field :book, Book, null: true do
      argument :id, ID, required: true
    end
    def book(id:)
      DistributedModel.new(id: id, model_name: "Book")
    end
  end
end

module Types
  class Book < BaseObject
    field :id, ID, null: false

    field :title, String, null: false
    def title
      book_details.title
    end

    field :author, Types::Author, null: false
    def author
      DistributedModel.new(id: book_details.author_id, model_name: "Author")
    end

    private

    def book_details
      BookService.get(id: object.id)
    end
  end
end

module Types
  class Author < BaseObject
    field :id, ID, null: false

    field :name, String, null: false
    def name
      author_details.name
    end

    field :books, [Types::Book], null: true
    def books
      book_ids = BookService.get_book_ids_by_author(author_id: object.id)

      book_ids.map do |book_id|
        DistributedModel.new(id: book_id, model_name: "Book")
      end
    end

    field :first_book, Types::Book, null: true
    def first_book
      book = BookService.get_first_book_by_author(author_id: object.id)
      DistributedModel.new(id: book.id, model_name: "Book")
    end

    field :most_recent_book, Types::Book, null: true
    def most_recent_book
      book = BookService.get_latest_book_by_author(author_id: object.id)
      DistributedModel.new(id: book.id, model_name: "Book")
    end

    field :top_selling_book, Types::Book, null: true
    def top_selling_book
      top_seller = BookSalesService.get_top_seller_by_author(author_id: object.id)
      DistributedModel.new(id: top_seller.book_id, model_name: "Book")
    end

    private

    def author_details
      AuthorService.get(id: object.id)
    end
  end
end
```

In this example, I assume that Books have exactly one Author.  If a Book in this system does not have an Author, I consider that an exceptional case and would be fine adding an error to the top-level errors array.

Okay, so, now I can make the following query and get back Book and Author data:

```
query GetBook($bookId: ID!) {
  book(id: $bookId) {
    id
    title
    author {
      id
      name
      firstBook {
        id
        title
      }
      topSellingBook {
        id
        title
      }
    }
  }
}
```

Now, if an error occurs, let's say that my `BookSalesService` returns a Book ID that my `BookService` knows nothing about, which results in a 404 error. Let's also assume that I am handling that execution error such that my GraphQL query result would look like:

```
{
  "data": {
    "book": {
      "id": "<id>",
      "name": "<name>",
      "firstBook": { ... },
      "topSellingBook": null
    }
  },
  "errors": [
    {
      "message": "BookService returned a 404 for the top-selling book",
      "path": ["book", "author", "topSellingBook", "title"]
    }
  ]
}
```

That is not terrible, but I was hoping that I could encode errors alongside my data in my schema.  In essence, I want my query and my errors to look something like this:

```
query GetBook($bookId: ID!) {
  book(id: $bookId) {
    id
    title
    author {
      id
      name
      firstBook {
        id
        title
      }
      topSellingBook {
        book {
          id
          title
        }
        errors {
          code
          message
        }
      }
    }
  }
}
```

The thing is, I am struggling to figure out how to achieve that result without possibly considering lookahead or eagerly fetching data that I may not need.

For example, I could create a new `TopSellingBook` type in my schema.

```ruby
# Change the Author type to use the TopSellingBook type
module Types
  class Author < BaseObject
    ...

    # Let the TopSellingBook type resolve its data
    field :top_selling_book, Types::TopSellingBook, null: true
    def top_selling_book
      { author_id: object.id }
    end
  end
end

module Types
  class TopSellingBook
    field :book, Types::Book, null: true
    def book
      book = book_details
      DistributedModel.new(id: book.id, model_name: "Book")
    rescue BookService::BookNotFound => e
      nil
    end

    field :errors, [Type::Error], null: false
    def errors
      book_details
      []
    rescue BookService::BookNotFound => e
      [Error.new(message: e.message, code: "book-not-found")]
    end

    private

    def top_seller
      BookSalesService.get_top_seller_by_author(author_id: object[:author_id])
    end

    def book_details
      BookService.get(id: top_seller.book_id)
    end
  end
end
```

So, given that implementation, I would be making a call to the `BookService` once in `TopSellingBook` and once in `Book`.  With a batch loader, I wouldn't be making those calls twice, though, so that seems solvable.  The piece that troubles me more, however, is, "What if it takes more than one service call to verify that the Book can be hydrated?"  I have also coupled my `TopSellingBook` and `Book` type implementations.  The `TopSellingBook` type needs to know what service calls `Book` might make to resolve itself, which doesn't seem ideal. This query, for instance, would make a needless call to the `BookService` unless we add lookahead logic in `TopSellingBook`.

```
query GetBook($bookId: ID!) {
  book(id: $bookId) {
    id
    title
    author {
      id
      name
      firstBook {
        id
        title
      }
      topSellingBook {
        book {
          id
        }
      }
    }
  }
}
```

For what it is worth, I have seen community resources suggest using union types to encode error cases in one's schema, but I am not sure that addresses my issue. In order to choose which type of a given union to return/resolve, my `TopSellingBook`, it seems to me, would still need to attempt to eagerly fetch `Book` data or factor in lookahead inspection logic.

For example, in the union case, we would need some additional types to define the union type:

```ruby
module Types
  class BookNotFound < BaseObject
    field :error, String, null: false
    def error
      object.message
    end
  end
end

module Types
  class BookUnion < BaseUnion
    description "Objects that may be returned for a Book query"
    possible_types Book, BookNotFound

    def self.resolve_type(object, context)
      if object.is_a?(BookService::BookNotFound)
        Types::BookNotFound
      else
        Types::Book
      end
    end
  end
end
```

Then we could implement our TopSellingBook type:

```ruby
module Types
  class TopSellingBook
    field :book, Types::BookUnion, null: true
    def book
      book = book_details
      DistributedModel.new(id: book.id, model_name: "Book")
    rescue BookService::BookNotFound => e
      e
    end

    private

    def top_seller
      BookSalesService.get_top_seller_by_author(author_id: object[:author_id])
    end

    def book_details
      BookService.get(id: top_seller.book_id)
    end
  end
end
```

The query would change to handle the different types in the union:

```
query GetBook($bookId: ID!) {
  book(id: $bookId) {
    id
    title
    author {
      id
      name
      firstBook {
        id
        title
      }
      topSellingBook {
        ... on Book {
          id
          title
        }

        ... on BookNotFound {
          error
        }
      }
    }
  }
}
```

But the `Types::TopSellingBook` still needs to eagerly load (or attempt to eagerly load) `book_details` to resolve its book field.

**Comment by rmosolgo** [comment]
Hey, thanks for the thorough write-up! I'm trying to understand your setup, let's see if this helps. You mention you'd _like_ to have a schema that works like this: 

```graphql 
      topSellingBook {
        book {
          id
          title
        }
        errors {
          code
          message
        }
      }
```

But the problem is, how do you reliably and efficiently populate `errors`? 

With mutations, it happens to be easy because there's just one synchronous field where you can `rescue` errors and add them to the result. 

But with Objects, it's really complicated. You mentioned the issue of multiple service calls to populate data. Besides that, what if `errors` should contain errors from multiple fields, for example: 

```
topSellingBook {
  book { ... } 
  retailers { ... } 
  errors { ... } 
}
```

In that case, `errors` might contain entries from _both_ `book` and `retailers`. You wouldn't want to load `retailers` if `retailers` isn't in the query, etc ... In that case, there might be some lookahead implementation that would work. 


It'd be nice if you could somehow "listen" inside `def errors` to see what fields were requested by the client, and build errors that way. But GraphQL-Ruby runs fields in top-to-bottom order, so it'd be hard to do runtime detection in a case like this: 

```
topSellingBook {
  errors { ... } # errors would get executed first, and wouldn't know about `book` or `retailers`
  book { ... } 
  retailers { ... } 
}
```

Tricky 😖 

------

Here's an alternative suggestion: what about post-processing the result from GraphQL-Ruby? For example, if you get: 

```ruby
{
  "data" => { ... } 
  "errors" => [
    {
      "message": "BookService returned a 404 for the top-selling book",
      "path": ["book", "author", "topSellingBook", "title"]
    }
  ]
}
```

Maybe you could take entries from `"errors"` and _move_ them into `"data"`, for example: 

```
# @return [void] modify `response` by moving any "errors" into "data" if their "path" points to an object with an "errors" array 
def move_errors_to_data(response)
  if response["errors"].present?
    response["errors"].reject! do |error|
      target = response["data"]
      error["path"].each do |part|
        if (next_target = target[part]) && next_target.is_a?(Hash)
          target = next_target 
        end 
        if target["errors"] && target["errors"].is_a?(Array)
          target["errors"] << error
          true # make `reject!` remove this error 
        else 
          false # return false so that the error stays in the top-level array 
        end 
    end 
  end 
end 
```

It's a bit of a hack, but it might help prototype. Honestly, I can't think of any decent implementation that doesn't work either _that_ way, or by some lookahead wizardry (🙀 ). If it works for you, we could consider adding a first-class API to graphql-ruby, something like: 

```ruby
raise GraphQL::ExecutionError(message, as_data: true)
```

Where GraphQL-Ruby would basically do _that_ ☝️ for you. (Inside the runtime, it'd be easier to, say, only populate fields which the client actually requests. My hot take above shoves the whole error into the result, regardless of the client's selections 😖 .)

Sorry there's nothing better built in.

**Comment by jdlubrano** [comment]
@rmosolgo 

Firstly, thank you so much for taking the time to consider my issue.  I greatly appreciate your prompt and thorough response.

---

> I'm trying to understand your setup

Your understanding seems spot on to me 😀 .  Thank you for taking a crack at solving the problem with the `move_errors_to_data` method.  At the very least, I can give that a shot.

> It'd be nice if you could somehow "listen" inside def errors to see what fields were requested by the client, and build errors that way.

That would be extremely nice, but as you say, I couldn't think of a reasonable implementation to achieve that sort of thing.  Especially if resolvers could be running in parallel or be a part of a federated graph...it quickly gets out of hand, or at least it seems to me.

Regarding first-class support within `graphql-ruby` itself, one of my original thoughts when I was experimenting with different approaches was that it might be nice to add `rescue_from` blocks inside of a given type.  So

```ruby
module Types
  class TopSellingBook < BaseObject
    def initialize
      super
      @errors = []
    end

    rescue_from(BookService::BookNotFound) do |err, ...|
      @errors << err
    end

    field :book, Types::Book, null: true
    def book
      book = book_details
      DistributedModel.new(id: book.id, model_name: "Book")
    rescue BookService::BookNotFound => e
      nil
    end

    field :errors, [Type::Error], null: false
    def errors
      @errors
    end

    private

    def top_seller
      BookSalesService.get_top_seller_by_author(author_id: object[:author_id])
    end

    def book_details
      BookService.get(id: top_seller.book_id)
    end
  end
end
```

I am guessing that kind of API is only achievable at the schema level because that allows any `GraphQL::ExecutionError`s to be inserted into the top-level errors array, though.  Or maybe the complexity arises with catching and adding errors somewhere other than the top-level errors array 🤔 .  In any case, I like the `rescue_from` API offered at the schema level and wondered whether that API could be reused at the type level, but that would almost certainly require support within the runtime.  In any case, I am guessing supporting something like that would be a significant level of effort.

Do you have any thoughts on the approach using union types above?  Moving errors in the result works for the errors array implementation, but that same approach could not be leveraged for returning a specific type that would be a part of a union, correct?  The only [concrete server-side implementation using union types to represent errors that I have found](https://blog.logrocket.com/handling-graphql-errors-like-a-champ-with-unions-and-interfaces/) does eagerly load data from a database, so perhaps that is the only realistic approach to achieve that pattern.

**Comment by rmosolgo** [comment]
I don't have plans to work further on this, so I'm going to close the issue. I hope you found something that works for you! If anyone wants to restart conversation on this topic, please open a new issue.
